### PR TITLE
compiler: Add optional pass for runtime stability check

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

File: devito/passes/iet/errors.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            if f in wmovs:
+                break
+        else:
+            continue
+
+        name = sregistry.make_name(prefix='energy')
+        energy = Symbol(name=name, dtype=f.dtype)
+
+        eqns = [Eq(energy, 0.0),
+                Inc(energy, Abs(Pow(f.subs(f.time_dim, 0), 2)))]
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/iet/errors.py:60]: OK

**Review by FabioLuporini — COMMENTED** [review]
OK I've tweaked things

new generated code:

```
    /* Stability check */
    if ((time)%(100) == 0)
    {
      int check0 = is_finite0(x_M,x_m,y_M,y_m,nthreads,u_vec);
      if (!check0)
      {
        return 100;
      }
    }
```

and

```
static int is_finite0(const int x_M, const int x_m, const int y_M, const int y_m, const int nthreads, struct dataobj *restrict u_vec)
{
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]]) u_vec->data;

  float accumulator = 0.0F;
  #pragma omp parallel num_threads(nthreads)
  {
    #pragma omp for collapse(2) schedule(static,1) reduction(+:accumulator)
    for (int x = x_m; x <= x_M; x += 1)
    {
      for (int y = y_m; y <= y_M; y += 1)
      {
        accumulator += u[0][x + 2][y + 2];
      }
    }
  }
  return isfinite(accumulator);
}
```

**Comment by FabioLuporini** [comment]
also added another tiny test

**Review by EdCaunt — APPROVED** [review]
Looks good

**Review by mloubout — APPROVED** [review]

File: devito/passes/iet/errors.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    # NOTE: Stability is a domain-specific concept, hence looking for time
+    # Iterations and TimeFunctions is acceptable
+    efuncs = []
+    includes = []
+    mapper = {}
+    for n in FindNodes(Iteration).visit(iet):
+        if not n.dim.is_Time:
+            continue
+
+        functions = [f for f in FindSymbols().visit(n)
```
Comment (mloubout) [inline devito/passes/iet/errors.py:42]: Nitpicking: Might be cases where there is no stepping dim and only time dime, but can be for later if someone runs into it (unlikely).

**Review by mloubout — APPROVED** [review]

**Pull request merged**